### PR TITLE
Fix issues with GDScript first-time parsing

### LINKED ISSUE #71570: Miscellaneous problems parsing GDScript in the scan thread
### Godot version

4.0-dev

### System information

all

### Issue description

There are a couple of problems in the scan thread when parsing script files (specifically GDScript).

The first one is relatively minimal:

https://github.com/godotengine/godot/blob/master/modules/gdscript/gdscript.cpp#L2468
This should be more tolerant to parse errors, so even if parse is not ok, this information will almost certainly still be there in the parse tree.

The second one is a design problem of the feature:
https://github.com/godotengine/godot/blob/master/editor/editor_file_system.cpp#L925
this has the following problems:
* This is way too early to load a script, most clases in the filesystem have not been scanned and it is highly likely the parsing will fail. this should be done later on in the scan scripts phase.
* The function to get documentation should still *not* load the script resource in a thread. This is dangerous as the script could have init logic not meant to run on a thread. This should be a dedicated function that uses the parser on the file on disk without opening the resource.

### Steps to reproduce

Here is a script that lets you generate a lot of Godot classes, you can see the second problem appear easily if you use this (I have no idea where this cache goes, but it seems the error only happens on the first time:
image

Python script: 
```Python
import random

class_count = 500

class_inherit_from = ["Sprite2D","ImageTexture","Control","ArrayMesh"]

for i in range(class_count):
    cname = "Class"+str(i)
    inh = ""
    if (i > 0 and i % 5 == 0):
        inh = "Class"+str(int(random.randrange(i)))
    else:
        inh = class_inherit_from[random.randrange(len(class_inherit_from))]
            
    cstr = "class_name "+cname+" extends "+inh    
    f = open(cname+".gd","w")
    f.write(cstr+"\n\nfunc _init():\n\tpass\n")
```

### Minimal reproduction project

Use script above in an empty project.

### PR DESCRIPTION
**Pull request by adamscott**
Fixes #71570

### tl;dr
This fixes the parse errors of GDScript files when fresh importing a project (or when running the project after deleting the `.godot` directory)

Removed the `err == OK` check in `String GDScriptLanguage::get_global_class_name()`. Doesn't crash when there's a parse error, but it doesn't really returns the global class name neither when there's a parse error. So, I don't really know if it's fixing something.

This is fixed by running `_scan_script_classes(p_dir)`. This parses files with `fi->script_class_name` set by `_get_global_script_class()` and it seems to do the trick. It creates global classes for those scripts, which future script loads will be able to refer to.

The scripts are still loaded in a thread, but after the part one fix. The init logic of the scripts are never ran, as only the source code is loaded. I don't see which code not meant to run on a thread could be run.

### Conclusion

I tested a fresh import of classes and no parse error were signalled. And the documentation was complete after the import.

### CODE REVIEW

**Comment by akien-mga** [comment]
There's a failing test:
```
===============================================================================
./modules/gdscript/tests/gdscript_test_runner_suite.h:43:
TEST SUITE: [Modules][GDScript]
TEST CASE:  Script compilation and runtime

modules/gdscript/tests/gdscript_test_runner.cpp:190: FATAL ERROR: An error occurred while generating class index.

===============================================================================
```

**Comment by adamscott** [comment]
Closing. Superseded by #71628.
